TELCODOCS-1279: First draft (#64784)
openshift/openshift-docs: modules/kmm-replacing-in-tree-modules-with-out-of-tree-modules.adoc

// Module included in the following assemblies:
//
// * hardware_enablement/kmm-kernel-module-management.adoc

:_content-type: CONCEPT
[id="kmm-replacing-in-tree-modules-with-out-of-tree-modules_{context}"]
= Replacing in-tree modules with out-of-tree modules

You can use Kernel Module Management (KMM) to build kernel modules that can be loaded or unloaded into the kernel on demand. These modules extend the functionality of the kernel without the need to reboot the system. Modules can be configured as built-in or dynamically loaded.

Dynamically loaded modules include in-tree modules and out-of-tree (OOT) modules. In-tree modules are internal to the Linux kernel tree, that is, they are already part of the kernel. Out-of-tree modules are external to the Linux kernel tree. They are generally written for development and testing purposes, such as testing the new version of a kernel module that is shipped in-tree, or to deal with incompatibilities.

> Some modules loaded by KMM could replace in-tree modules already loaded on the node. To unload an in-tree module before loading your module, set the `.spec.moduleLoader.container.inTreeModuleToRemove`. This is an example for module replacement for all kernel mappings:

Author: Replaced with "Some modules loaded by KMM could replace in-tree modules already loaded on the node. To unload an in-tree module before loading your module, set the `.spec.moduleLoader.container.inTreeModuleToRemove` field. The following is an example for module replacement for all kernel mappings:

> 

Reviewer: I think we'll want to include the `# ...` ellipses since this is a snippet: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#yaml-formatting-for-kubernetes-and-openshift-api-objects

Author: Added the `# ...` snippet.

[source,yaml]
----
spec:
  moduleLoader:
    container:
      modprobe:

>         moduleName: mod_a

Author: From the style guide: Alternatively, if you choose to use an example value instead, do not italicize the example value and do not place it in angle brackets.

>       inTreeModuleToRemove: mod_b

Author: See response above.

----

> In this example, the ModuleLoader pod uses `inTreeModuleToRemove` to unload the in-tree `mod_b` before loading `mod_a`

Reviewer 2: @StephenJamesSmith same question. If it is, it needs backticks.

> In this example, the ModuleLoader pod uses `inTreeModuleToRemove` to unload the in-tree `mod_b` before loading `mod_a`

Author: @JoeAldinger Added backticks.

from the ModuleLoader image.

> When the ModuleLoader pod is terminated and `mod_a` is unloaded, `mod_b` will not be loaded again.

Author: Changed to "`mod_b` is not be loaded again."

> This is an example for module replacement for specific kernel mappings:

Reviewer: Maybe consider: 

```suggestion
The following is an example for module replacement for specific kernel mappings:
```

Author: Made this change.

[source,yaml]
----

> spec:

Author: @cdvultur Whoops. Sorry. Fixed.

> spec:

Reviewer: I think we'll want to include the `# ...` ellipses since this is a snippet: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#yaml-formatting-for-kubernetes-and-openshift-api-objects

Author: Added the `# ...` snippet.

>     moduleLoader:

Reviewer 3: Can we use the same layout as in the above example ? I see 2 spaces in the previous example and in here I see 4 spaces. 
It would be nice to have the same spacing.

Author: @cdvultur changed.

    container:
      kernelMappings:
        - literal: 6.0.15-300.fc37.x86_64
          containerImage: some.registry/org/my-kmod:6.0.15-300.fc37.x86_64

>           inTreeModuleToRemove: <module-name>

Reviewer: Suggestion from https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values

```suggestion
          inTreeModuleToRemove: <module_name>
```

Author: Made this change.

:_content-type: CONCEPT
[id="kmm-replacing-in-tree-modules-with-out-of-tree-modules_{context}"]
= Replacing in-tree modules with out-of-tree modules

You can use Kernel Module Management (KMM) to build kernel modules that can be loaded or unloaded into the kernel on demand. These modules extend the functionality of the kernel without the need to reboot the system.
Modules can be configured as built-in or dynamically loaded.

Dynamically loaded modules include in-tree modules and out-of-tree (OOT) modules.
In-tree modules are internal to the Linux kernel tree, that is, they are already part of the kernel.
Out-of-tree modules are external to the Linux kernel tree. They are generally written for
development and testing purposes, such as testing the new version of a kernel module that is
shipped in-tree, or to deal with incompatibilities.

Some modules loaded by KMM could replace in-tree modules already loaded on the node.
To unload an in-tree module before loading your module,
set the `.spec.moduleLoader.container.inTreeModuleToRemove`. This is an example for module replacement
for all kernel mappings:

[source,yaml]
----
spec:
  moduleLoader:
    container:

>       modprobe:

Author: @qbarrand Do we need to show usage of `modprobe -r <module_name>` and `modprobe -d <module_name>`?  Any other changes to these examples?

Reviewer 4: I believe we do not need to include that.

Reviewer 3: Agree with @qbarrand on this one, This should not be included in the docs. 
If a customer really wants it, he can look at the code.

Reviewer 4: I think the YAML example is fine. I see no reason to reference or explain the `modprobe` usage in this section.

Author: @qbarrand To be clear should I also remove `moduleName: mod_a`? 
![image](https://github.com/openshift/openshift-docs/assets/5238708/2decac3b-b02a-41ea-b64f-0d63eb4f3412)

        moduleName: mod_a

      inTreeModuleToRemove: mod_b
----

> In this example, The ModuleLoader pod uses `inTreeModuleToRemove` to unload the in-tree `mod_b` before loading `mod_a`

Reviewer 4: ```suggestion
In this example, the ModuleLoader pod uses `inTreeModuleToRemove` to unload the in-tree `mod_b` before loading `mod_a`
```

Reviewer 3: +1

Author: fixed.
